File: Algorithms/116.PopulatingNextRightPointersInEachNode.java
public class Solution {
    public void connect(TreeLinkNode root) {
        if (root == null) {
            return ;
        }
        while (root != null) {
            TreeLinkNode curr_node = root;
            while (curr_node != null) {
                if (curr_node.left != null) 
                    curr_node.left.next = curr_node.right;
                if (curr_node.right != null) {
                    curr_node.right.next = curr_node.next != null ? curr_node.next.left : null;
                }
                curr_node = curr_node.next;
            }
            root = root.left;
        }
    }
    /*
    // not right
    public void connect(TreeLinkNode root) {
        if (root == null) {
            return ;
        }
        Queue<TreeLinkNode> q = new LinkedList<>();
        Queue<TreeLinkNode> next_level_q = new LinkedList<>();
        TreeLinkNode previous = null;
        q.add(root);
        while (q.size() != 0) {
            TreeLinkNode p = q.remove();
            if (previous != null) {
                previous.next = p;
            }
            previous = p;
            if (p.left != null) {
                next_level_q.add(p.left);
            }
            if (p.right != null) {
                next_level_q.add(p.right);
            }
            if (q.size() == 0) {
                p.next = null;
                q = next_level_q;
                next_level_q = new LinkedList<>();
                previous = null;
            }
        }
    }
    */
}

File: Algorithms/cpp/116.PopulatingNextRightPointersInEachNode.cpp
class Solution {
public:
    void connect(TreeLinkNode *root) {
        if (root == NULL) {
            return ;
        }
        while (root != NULL) {
            TreeLinkNode *curr_node = root;
            while (curr_node != NULL) {
                if (curr_node->left != NULL) 
                    curr_node->left->next = curr_node->right;
                if (curr_node->right != NULL) {
                    curr_node->right->next = curr_node->next != NULL ? curr_node->next->left : NULL;
                }
                curr_node = curr_node->next;
            }
            root = root->left;
        }
    }
    /*
    // not right
    void connect(TreeLinkNode *root) {
        if (root == NULL) {
            return ;
        }
        queue<TreeLinkNode*> q;
        queue<TreeLinkNode*> next_level_q;
        TreeLinkNode *previous = NULL;
        q.push(root);
        while (!q.empty()) {
            TreeLinkNode *p = q.front();
            if (previous != NULL) {
                previous->next = p;
            }
            previous = p;
            q.pop();
            if (p->left != NULL) {
                next_level_q.push(p->left);
            }
            if (p->right != NULL) {
                next_level_q.push(p->right);
            }
            if (q.empty()) {
                p->next = NULL;
                q.swap(next_level_q);
                previous = NULL;
            }
        }
    }
    */
};
